Add CIBW_BEFORE_TEST to customize test environment (#242)
joerick/cibuildwheel: docs/options.md

CIBW_TEST_EXTRAS: test,qt
```

### `CIBW_BEFORE_TEST` {: #before-test}

> > Execute a shell command preparing test environment

Reviewer: ```suggestion
> Execute a shell command before testing each wheel
```

> A shell command to run after setup test environment. This option allows you to run a command in **each** Python environment before the `pip wheel` command. This is useful if you need to install non pip package, change values of environment variables

Reviewer: ```suggestion
A shell command to run in **each** test virtual environment, before your wheel is installed and tested. This is useful if you need to install non pip package, change values of environment variables
```

> or perform multi step pip installation (like install `scikit-build` or `cython` before install test package)

Reviewer: These reasons are a little curious, actually.

- Non-pip packages don't care about the virtualenv, so can be installed in BEFORE_BUILD. 
- Changing environment variables won't work, since the command is run in a subshell
- Maybe I'm missing something, but I can't see the need to install anything _before_ installing the wheel? Wheel installation is just an unzip and move, no code within the package is run. So why is there a need to install dependencies before that? It seems that TEST_REQUIRES would be sufficient for these cases.

Author: Example situation:

 - PyQt5 from 5.14.0 has wheel with `manylinux2014` platform tag. Earlier are published with `manylinux1`. `manylinux2014` needs pip  in version 19.3. Example solution with `BEFORE_BUILD` is to manually install `PyQt5==5.13.2`
- [cmake](https://pypi.org/project/cmake/) package does not provide wheels for x86 python 3.8. So only option is to build it from source. But cmake package needs `scikit-build` for build and this dependency needs to be importable (installed in previous transatction). And package [blosc](https://pypi.org/project/blosc/) does not provide wheel and need cmake package to be installed. And tests need `blosc` package. 
- i do not have given example but it is possible to have similar story with `Cython` 

Also non pip package can care about virtualenv (location of python executable). Here example: https://github.com/LLNL/zfp

And change environment variable I think about `ENV1=1 ENV2=7 command`

> If dependencies are required to build your wheel (for example if you include a header from a Python module), set this to `pip install .`, and the dependencies will be installed automatically by pip. However, this means your package will be built twice - if your package takes a long time to build, you might wish to manually list the dependencies here instead.

Reviewer: ```suggestion
```

The active Python binary can be accessed using `python`, and pip with `pip`; `cibuildwheel` makes sure the right version of Python and pip will be executed. `{project}` can be used as a placeholder for the absolute path to the project's root and will be replaced by `cibuildwheel`.

On Linux and macOS, the command is run in a shell, so you can write things like `cmd1 && cmd2`.

Platform-specific variants also available:<br/>
 `CIBW_BEFORE_TEST_MACOS` | `CIBW_BEFORE_TEST_WINDOWS` | `CIBW_BEFORE_TEST_LINUX`

#### Examples
```yaml
# install packages needed to build test dependencies

> CIBW_BEFORE_TEST: pip install cmake scikit-build

Reviewer: See above, I'm not sure why these aren't TEST_REQUIRES

Author: As I mentioned before. Some dependencies need to be importable in `setup.py`. And still some authors of packages does not provide wheels or it takes over month since release of python 3.8 to wheel for this python. And such things allow maintainer to have full test case without waiting on other persons. 

```
File "/tmp/pip-install-f9rmc9hk/cmake/setup.py", line 7, in <module>
          from skbuild import setup
      ModuleNotFoundError: No module named 'skbuild'
```

Reviewer 2: @joerick I know `setup_requires` doesn't always work, for some reason. So that would be good.

But indeed, this might not be the perfect example to put in the docs, either?

# install test dependencies with overwritten environment variables. 

> CIBW_BEFORE_TEST: CC=gcc CXX=g++ pip install -r requirements.txt

Reviewer: Curious - why would this be necessary?

Author: My case is that i want to build package with OMP support for MacOS. MacOS clang does not support this. So I need to use brew g++ for this. But one of test dependencies need to be compiled with system `clang`. Otherwise I got such message:

```
python(53258,0x7fff9481d380) malloc: *** error for object 0x10f526c00: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug
```

Other case which I can imagine is that my package is build against newest version of some library. And test dependency is build against older (system one). And I need to change `LD_LIBRARY_PATH` etc to avoid collision.

Reviewer 2: This confusion makes me wonder; wouldn't a different example be better for the docs, then?

Author: What do you think about 
`CFLAGS="-I ... -L ..." pip install package name` 
for custom library location?

Reviewer 2: Huh, maybe. But .. that's again something where you'd ask the question where a user would ask why this is not `CIBW_TEST_ENVIRONMENT`...

Author: Totally different example. Install non pip package like: https://github.com/LLNL/zfp

For this...  (original example)
MacOS system clang does not support OpenMP. So if package needs OpenMP support then there is need to compile it with gcc/g++ (ex from brew). 

For second. I meet problem when test for one package fail on Azure Pipeline but not fail on my computer. And this fail was connected with package used in test and build from source. And for me possibility to control compilation of single package is big profit.

# chain commands using &&

> CIBW_BEFORE_TEST: yum install -y libffi-dev && pip install .

Reviewer: `pip install .` would be an anti-pattern, since it would build and install from code, rather than from the wheel.

Author: Agree. I copy it from `CIBW_BEFORE_BUILD` documentation. Whole example is bad. Installig things system side during testis also bad idea. Maybe `pip install scikit-build && pip install cmake`?

Author: yes
